File: apps/web/lib/bom-validation.ts
/**
 * apps/web/lib/bom-validation.ts — TS mirror of `packages/services/m1_baseline/bom_validation.py`.
 *
 * Story 2.2 — Task 5.5 mirror (drift-checked by `tests/integration/test_bom_validation_consistency.py`).
 *
 * Pure functions only — no DOM, no fetch, no clock. The matrix UI calls
 * these for the live "is_complete" check + "missing %" toast.
 *
 * Cross-language parity with Python:
 * - TARGET_TOTAL === Decimal("100.0000")
 * - sumRatios quantizes to 4 decimal places (ROUND_HALF_EVEN equivalent)
 * - isCompleteBom === sumRatios(...) === TARGET_TOTAL
 *
 * Uses `decimal.js` for arithmetic (AD-8 deferred — money paths use it;
 * ratio follows the same rule).
 */

import Decimal from "decimal.js";

// Configure Decimal for AD-8 + Story 0.4 chunk-B parity:
// ROUND_HALF_EVEN (= banker's rounding) for quantization.
Decimal.set({ rounding: Decimal.ROUND_HALF_EVEN });

/** Target total — 100.0000% (NUMERIC(7,4) precision). */
export const TARGET_TOTAL = new Decimal("100.0000");

/** Quantum — 4 decimal places. */
const QUANTUM = new Decimal("0.0001");

/**
 * Sum an iterable of ratios as Decimal with 4-decimal precision.
 *
 * Accepts `number`, `string`, or `Decimal`. Strings are parsed via
 * `Decimal(value)` so `"12.3456"` works. Non-numeric inputs throw.
 */
export function sumRatios(
  rows: Iterable<number | string | Decimal>,
): Decimal {
  let total = new Decimal(0);
  for (const r of rows) {
    if (r instanceof Decimal) {
      total = total.plus(r);
    } else if (typeof r === "number") {
      // Decimal() traps NaN / Infinity.
      total = total.plus(new Decimal(r));
    } else if (typeof r === "string") {
      total = total.plus(new Decimal(r));
    } else {
      throw new TypeError(
        `ratio must be Decimal, number, or string; got ${typeof r}`,
      );
    }
  }
  return total.toDecimalPlaces(4, Decimal.ROUND_HALF_EVEN);
}

/**
 * Is the BOM complete? (A6 axiom: `sum === 100.0000`).
 *
 * Empty BOM is NOT complete. Over-100 is NOT complete. Only exactly
 * 100.0000 is complete.
 */
export function isCompleteBom(
  rows: Iterable<number | string | Decimal>,
): boolean {
  return sumRatios(rows).equals(TARGET_TOTAL);
}

/**
 * How much is needed to reach 100.0000%? Clamped at zero.
 */
export function missingToComplete(
  rows: Iterable<number | string | Decimal>,
): Decimal {
  const total = sumRatios(rows);
  const delta = TARGET_TOTAL.minus(total);
  if (delta.lessThanOrEqualTo(0)) return new Decimal("0.0000");
  return delta.toDecimalPlaces(4, Decimal.ROUND_HALF_EVEN);
}

/**
 * Quantize a single ratio to 4 decimal places (ROUND_HALF_EVEN).
 *
 * Defense-in-depth — Pydantic should have caught invalid ratios at the
 * wire boundary. This is the last layer before the API call.
 */
export function quantizeRatio(value: Decimal | number | string): Decimal {
  if (value instanceof Decimal) {
    return value.toDecimalPlaces(4, Decimal.ROUND_HALF_EVEN);
  }
  return new Decimal(value).toDecimalPlaces(4, Decimal.ROUND_HALF_EVEN);
}

// ── BOM type rules (PRD §6.1) ───────────────────────────────────
// Mirrors `packages/services/m1_baseline/schemas.py::BOMParentType` / `BOMChildType`.

export const BOMParentTypes: ReadonlySet<ProductType> = new Set([
  "product",
  "semi_product",
]);

export const BOMChildTypes: ReadonlySet<ProductType> = new Set([
  "material",
  "semi_product",
]);

export function isValidBomParent(productType: ProductType): boolean {
  return BOMParentTypes.has(productType);
}

export function isValidBomChild(productType: ProductType): boolean {
  return BOMChildTypes.has(productType);
}

// We intentionally do NOT import ProductType to keep this module free of
// circular imports; the TS ProductType literal union lives in
// `apps/web/lib/api-client.ts`. The mirror test only checks string
// equality, so the union membership is implicit.
type ProductType =
  | "product"
  | "semi_product"
  | "material"
  | "goods"
  | "service";

File: packages/services/m1_baseline/bom_validation.py
"""packages.services.m1_baseline.bom_validation — pure BOM validators (Story 2.2).

Pure-Python, stdlib-only module (AD-1 / AD-5). NO DB, NO clock, NO random.

Holds the **pure** transformations and invariants for the BOM matrix:

- `TARGET_TOTAL` — the 100% invariant (AD-8 + A6 axiom).
- `sum_ratios` — Decimal-arithmetic sum, no float (AD-8 strict).
- `is_complete_bom` — `sum == TARGET_TOTAL` check (A6 derived invariant).
- `missing_to_complete` — `max(TARGET_TOTAL - sum, 0)` (UX: "{missing}% 부족").
- `quantize_ratio` — `ROUND_HALF_EVEN` to 4 decimal places (AD-8 parity).

The DB-level `bom_lines.ratio` is `NUMERIC(7,4)` (max 100.0000, 4 decimals).
The Pydantic schema `BOMRowInput.ratio` carries `max_digits=7, decimal_places=4`
defense-in-depth at the wire boundary. This module is the third layer of
defense: any caller that constructs a `Decimal` for storage must run it
through `quantize_ratio()` first.

Why this lives in `packages/services/` (not `apps/api/`):
- Story 2.2 TS mirror lives at `apps/web/lib/bom-validation.ts`.
- Drift is caught by `tests/integration/test_bom_validation_consistency.py`.
- AD-11 layer rule — `apps/api/core/` may NOT import `packages.cost_engine`
  directly, and pure helpers belong in `packages/services/` so both API
  and web can share them.
"""

from __future__ import annotations

from collections.abc import Iterable
from decimal import ROUND_HALF_EVEN, Decimal
from typing import Final

# ── A6 invariant — 100.0000% complete ───────────────────────────
# NUMERIC(7,4) → 4 decimal places, max 100.0000. Matches the DB column
# precision so the on-disk sum and the in-Python sum are bit-identical.
TARGET_TOTAL: Final[Decimal] = Decimal("100.0000")

# Quantization granularity — `Decimal("0.0001")` enforces 4-decimal-place
# rounding via `ROUND_HALF_EVEN` (AD-8 + Story 0.4 chunk-B Decimal.set parity).
_RATIO_QUANTUM: Final[Decimal] = Decimal("0.0001")


# ── Typed errors (mapped to HTTP by handlers.py) ──────────────
class BOMValidationError(ValueError):
    """Base class for BOM pure-validation failures.

    Subclasses carry `details` and `trace_id` for the AD-15 §4 envelope.
    """


class InvalidRatioTypeError(BOMValidationError):
    """422 BOM_INVALID_RATIO — a non-numeric value was passed to `sum_ratios`.

    Defense-in-depth — Pydantic `BOMRowInput.ratio: Decimal = Field(...)`
    should reject this at the wire boundary. If a service-level call
    skips Pydantic, this exception fires instead of silently coercing
    (which would invite `float` arithmetic drift).
    """

    def __init__(self, value: object) -> None:
        super().__init__(f"ratio must be Decimal, int, or float; got {type(value).__name__}")
        self.value = value


# ── Public helpers ────────────────────────────────────────────
def sum_ratios(rows: Iterable[Decimal | float | int]) -> Decimal:
    """Pure: sum an iterable of ratios as Decimal with 4-decimal precision.

    Defense-in-depth: rejects non-numeric inputs (would otherwise coerce
    to Decimal incorrectly or raise `TypeError` from `Decimal(float)`'s
    own checks).

    Args:
        rows: An iterable of `Decimal`, `int`, or `float`. `float` is
            accepted (not encouraged) for legacy callers; the function
            uses `Decimal(value)` conversion which traps `NaN` / `Inf`.

    Returns:
        `Decimal` quantized to 4 decimal places (the NUMERIC(7,4) DB
        precision). `ROUND_HALF_EVEN` per AD-8.

    Raises:
        InvalidRatioTypeError: A non-numeric value was in the iterable.
        decimal.InvalidOperation: `NaN` / `Inf` (from float → Decimal).
    """
    total = Decimal("0")
    for r in rows:
        if isinstance(r, Decimal | int):
            total += Decimal(r)
        elif isinstance(r, float):
            # Decimal(float) traps NaN / Inf via InvalidOperation.
            total += Decimal(r)
        else:
            raise InvalidRatioTypeError(r)
    return total.quantize(_RATIO_QUANTUM, rounding=ROUND_HALF_EVEN)


def is_complete_bom(rows: Iterable[Decimal]) -> bool:
    """Pure: is the BOM complete? (A6 axiom: `sum == 100.0000`).

    An empty BOM is NOT complete (`sum == 0`, expected `100`).
    A BOM summing to `120` is NOT complete (over-100 is invalid).
    A BOM summing to `99.9999` is NOT complete (off-by-tiny).
    Only `100.0000` is complete.

    Args:
        rows: An iterable of ratio `Decimal`s.

    Returns:
        `True` iff the sum is exactly `100.0000`.
    """
    return sum_ratios(rows) == TARGET_TOTAL


def missing_to_complete(rows: Iterable[Decimal]) -> Decimal:
    """Pure: how much is needed to reach 100.0000%?

    Clamped at zero (over-100 still reports `0` so the UX toast says
    "비중 합 100% 필요 (현재 X%)" instead of "비중 합 100% 필요 (-20% 부족)").
    Used by the matrix UI to render "{missing}% 부족" inline.

    Args:
        rows: An iterable of ratio `Decimal`s.

    Returns:
        `Decimal` quantized to 4 decimal places in `[0, 100.0000]`.
    """
    total = sum_ratios(rows)
    delta = TARGET_TOTAL - total
    if delta <= Decimal("0"):
        return Decimal("0.0000")
    return delta.quantize(_RATIO_QUANTUM, rounding=ROUND_HALF_EVEN)


def quantize_ratio(value: Decimal) -> Decimal:
    """Pure: round a single ratio to 4-decimal NUMERIC(7,4) precision.

    `ROUND_HALF_EVEN` per AD-8 / Story 0.4 chunk-B Decimal.set parity.

    Args:
        value: A `Decimal` ratio.

    Returns:
        `Decimal` quantized to 4 decimal places.

    Raises:
        InvalidRatioTypeError: Non-Decimal / non-numeric input.
    """
    if not isinstance(value, Decimal | int):
        raise InvalidRatioTypeError(value)
    return Decimal(value).quantize(_RATIO_QUANTUM, rounding=ROUND_HALF_EVEN)


# Re-exports — keep the public API flat.
__all__ = [
    "BOMValidationError",
    "InvalidRatioTypeError",
    "TARGET_TOTAL",
    "sum_ratios",
    "is_complete_bom",
    "missing_to_complete",
    "quantize_ratio",
]
